libraries/Microsoft.Bot.Builder.Skills/Dialogs/SkillDialog.cs

```diff
@@ -0,0 +1,118 @@
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Skills;
+using Microsoft.Bot.Builder.Skills.Dialogs;
+using Microsoft.Bot.Builder.TraceExtensions;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    /// <summary>
+    /// A sample dialog that can wrap remote calls to a skill.
+    /// </summary>
+    /// <remarks>
+    /// The options parameter in <see cref="BeginDialogAsync"/> must be a <see cref="SkillDialogArgs"/> instance
+    /// with the initial parameters for the dialog.
+    /// </remarks>
+    public class SkillDialog : Dialog
+    {
+        private readonly IStatePropertyAccessor<BotFrameworkSkill> _activeSkillProperty;
+        private readonly ConversationState _conversationState;
+        private readonly SkillDialogOptions _dialogOptions;
+
+        public SkillDialog(SkillDialogOptions dialogOptions, ConversationState conversationState)
+            : base(nameof(SkillDialog))
+        {
+            _dialogOptions = dialogOptions ?? throw new ArgumentNullException(nameof(dialogOptions));
+            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
+            _activeSkillProperty = conversationState.CreateProperty<BotFrameworkSkill>($"{typeof(SkillDialog).FullName}.ActiveSkillProperty");
+        }
+
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default)
+        {
+            if (!(options is SkillDialogArgs dialogArgs))
+            {
+                throw new ArgumentNullException(nameof(options), $"Unable to cast {nameof(options)} to {nameof(SkillDialogArgs)}");
+            }
+
+            // Store Skill information for this dialog instance
+            await _activeSkillProperty.SetAsync(dc.Context, dialogArgs.Skill, cancellationToken).ConfigureAwait(false);
+
+            await dc.Context.TraceActivityAsync($"{GetType().Name}.BeginDialogAsync()", label: $"Using activity of type: {dialogArgs.ActivityType}", cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            Activity skillActivity;
+            switch (dialogArgs.ActivityType)
+            {
+                case ActivityTypes.Event:
+                    var eventActivity = Activity.CreateEventActivity();
+                    eventActivity.Name = dialogArgs.Name;
+                    skillActivity = (Activity)eventActivity;
+                    break;
+
+                case ActivityTypes.Message:
+                    var messageActivity = Activity.CreateMessageActivity();
+                    messageActivity.Text = dialogArgs.Text;
+                    skillActivity = (Activity)messageActivity;
+                    break;
+
+                default:
+                    throw new ArgumentException($"Invalid activity type in {dialogArgs.ActivityType} in {nameof(SkillDialogArgs)}");
```

Comment:
Little weird to see an argument exception so far down a function - perhaps check args at the top.

Perhaps put the switch in its own function - maybe that's excessive - maybe just use an if?

Reply:
Moved validation code into a separate helper function and updated it to throw different types of exception bases on missing arguments and properties.